lib/elixir/lib/kernel.ex

```diff
@@ -5462,6 +5462,89 @@ defmodule Kernel do
     raise ArgumentError, "modifier must be one of: s, a, c"
   end
 
+  @doc """
+  Handles the `~PID` sigil.
+
+  This function allows creating arbitrary PIDs, even the ones that don't actually exist in the VM.
+  As such, it should only be used for debugging purposes.
+
+  ## Examples
+
+      iex> ~PID<0.108.0>
+      ~PID<0.108.0>
+
+  """
+  @doc since: "1.11.0"
+  def sigil_PID(string, [] = _modifiers) do
+    :erlang.list_to_pid('<#{string}>')
+  end
+
+  @doc """
+  Handles the `~Port` sigil.
+
+  This function allows creating arbitrary ports, even the ones that don't actually exist in the VM.
+  As such, it should only be used for debugging purposes.
+
+  ## Examples
+
+      iex> ~Port<0.6>
+      ~Port<0.6>
+
+  """
+  @doc since: "1.11.0"
+  def sigil_Port(string, [] = _modifiers) do
+    :erlang.list_to_port('#Port<#{string}>')
+  end
+
+  @doc """
+  Handles the `~Reference` sigil.
+
+  This function allows creating arbitrary references, even the ones that don't actually exist in the VM.
+  As such, it should only be used for debugging purposes.
+
+  ## Examples
+
+      iex> ~Reference<0.2283498464.2022703108.246828>
+      ~Reference<0.2283498464.2022703108.246828>
+
+  """
+  @doc since: "1.11.0"
+  def sigil_Reference(string, [] = _modifiers) do
+    :erlang.list_to_ref('#Ref<#{string}>')
+  end
+
+  @doc """
+  Handles the `~URI` sigil.
+
+  See `URI.parse/1` for more information.
+
+  ## Examples
+
+      iex> ~URI<https://elixir-lang.org>
+      ~URI<https://elixir-lang.org>
+
+  """
+  @doc since: "1.11.0"
+  defmacro sigil_URI({:<<>>, _, [string]} = _uri_string, [] = _modifiers) do
+    Macro.escape(URI.parse(string))
+  end
+
+  @doc """
+  Handles the `~Version` sigil.
+
+  See `Version.parse!/1` for more information.
+
+  ## Examples
+
+      iex> ~Version<2.0.1-alpha1>
+      ~Version<2.0.1-alpha1>
+
+  """
+  @doc since: "1.11.0"
+  defmacro sigil_Version({:<<>>, _, [string]} = _version_string, [] = _modifiers) do
+    Macro.escape(Version.parse!(string))
+  end
```

Comment:
So I have thought more about having this in Kernel vs in the respective modules, at least for non built-in types. The "issue" is that we already have many built-in types with sigils in Kernel, so I think it is best to keep all of them there, otherwise the line is arbitrary.

The other issue with the current implementation is that it is not "plug and play". For example, if something returns `~Decimal<1.0>`, i cannot simply use that in my code, unless the sigil is imported. This UI can be confusing. So my suggestion would be for us to, when failing to compiling due to a non-existing local function, we could check if:

1. If the function is a sigil starting with uppercase

2. If so, if there is a module with the name of the sigil and that exports a `sigil_NAME` function or macro. If so, the error message will add: "Perhaps you meant to "import Decimal, only: :sigils" before using it?" or similar

Thoughts?

Reply:
Yeah I think handling the error better at the compiler level is the way to go. In this case, I think it makes the case for having `~Version` and `~URI` out of Kernel even stronger.

Reply:
Let's separate these discussion. My main concern is: can you explain why ~Version is not in Kernel but ~Port is? Why is ~r in Kernel and not ~URI?

There is one heuristic I could come up with to explain it but it is so long that I am lazy of even typing it. :P

Reply:
Ahaha, okay, let's have that as a separate discussion, good point!

Reply:
+1 on error message from me.

Reply:
`~Port` is a built-in data type. `~Version` is not. We have `is_port` in `Kernel`, we don't have `is_version` in `Kernel`. I see a similar analogy with sigils.

Reply:
Yeah! It doesn't explain why we have sigils for Regex or Date/Time though. :/

So the rule is like "it has to be a built-in type and use a multi-letter sigil" and that relies on users knowing what is a built-in type and there are probably some gotchas still. I think it is easier to go full-Oprah on this and say "you get a Kernel sigil", "you get a Kernel sigil" and "you get a Kernel sigil".